src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java

```diff
@@ -2538,4 +2298,369 @@ public static void shutdownBlocking(long timeout, TimeUnit unit) throws Interrup
+
+
+            // we read the positions in a BRAF so we don't have to worry about an entry spanning a mmap boundary.
+            try (RandomAccessReader primaryIndex = RandomAccessReader.open(new File(descriptor.filenameFor(Component.PRIMARY_INDEX))))
+            {
+                long indexSize = primaryIndex.length();
+                long histogramCount = statsMetadata.estimatedPartitionSize.count();
+                long estimatedKeys = histogramCount > 0 && !statsMetadata.estimatedPartitionSize.isOverflowed()
+                                     ? histogramCount
+                                     : estimateRowsFromIndex(primaryIndex, descriptor); // statistics is supposed to be optional
+
+                if (recreateBloomFilter)
+                    bf = FilterFactory.getFilter(estimatedKeys, metadata.params.bloomFilterFpChance);
+
+                try (IndexSummaryBuilder summaryBuilder = summaryLoaded ? null : new IndexSummaryBuilder(estimatedKeys, metadata.params.minIndexInterval, Downsampling.BASE_SAMPLING_LEVEL))
+                {
+                    long indexPosition;
+
+                    while ((indexPosition = primaryIndex.getFilePointer()) != indexSize)
+                    {
+                        ByteBuffer key = ByteBufferUtil.readWithShortLength(primaryIndex);
+                        RowIndexEntry.Serializer.skip(primaryIndex, descriptor.version);
+                        DecoratedKey decoratedKey = metadata.partitioner.decorateKey(key);
+                        if (first == null)
+                            first = decoratedKey;
+                        last = decoratedKey;
+
+                        if (recreateBloomFilter)
+                            bf.add(decoratedKey);
+
+                        // if summary was already read from disk we don't want to re-populate it using primary index
+                        if (!summaryLoaded)
+                        {
+                            summaryBuilder.maybeAddEntry(decoratedKey, indexPosition);
+                        }
+                    }
+
+                    if (!summaryLoaded)
+                        summary = summaryBuilder.build(metadata.partitioner);
+                }
+            }
+
+            first = getMinimalKey(first);
+            last = getMinimalKey(last);
+        }
+
+        /**
+         * Load bloom filter from Filter.db file.
+         *
+         * @throws IOException
+         */
+        IFilter loadBloomFilter() throws IOException
+        {
+            try (DataInputStream stream = new DataInputStream(new BufferedInputStream(Files.newInputStream(Paths.get(descriptor.filenameFor(Component.FILTER))))))
+            {
+                return BloomFilterSerializer.deserialize(stream, descriptor.version.hasOldBfFormat());
+            }
+        }
+
+        public abstract SSTableReader build(Set<Component> components, StatsMetadata statsMetadata, OpenReason reason, SerializationHeader header);
```

Comment:
Fun Fact: Everything would compile if we removed this abstract method ;)

Reply:
because it uses the concrete type..